sdk/python/feast/feature_store.py

```diff
@@ -2173,26 +2194,44 @@ def _augment_response_with_on_demand_transforms(
                 )
 
         initial_response = OnlineResponse(online_features_response)
-        initial_response_df = initial_response.to_df()
+
+        # RB: lazy populate df and dict responses
+        initial_response_df: Optional[pd.DataFrame] = None
+        initial_response_dict: Optional[Dict[str, List[Any]]] = None
 
         # Apply on demand transformations and augment the result rows
         odfv_result_names = set()
         for odfv_name, _feature_refs in odfv_feature_refs.items():
             odfv = requested_odfv_map[odfv_name]
-            transformed_features_df = odfv.get_transformed_features_df(
-                initial_response_df,
-                full_feature_names,
-            )
+            if odfv.mode == "python":
+                if initial_response_dict is None:
+                    initial_response_dict = initial_response.to_dict()
+                transformed_features = odfv.get_transformed_features_dict(
+                    initial_response_dict,
+                    full_feature_names
+                )
+                feature_names = transformed_features
+            elif odfv.mode == "pandas":
+                if initial_response_df is None:
+                    initial_response_df = initial_response.to_df()
+                transformed_features = odfv.get_transformed_features_df(
+                    initial_response_df,
+                    full_feature_names
+                )
+                feature_names = transformed_features.columns
+            else:
+                raise Exception(f'Invalid OnDemandFeatureMode: {odfv.mode}. Expected one of "pandas" or "python".')
+
             selected_subset = [
-                f for f in transformed_features_df.columns if f in _feature_refs
+                f for f in feature_names if f in _feature_refs
             ]
 
-            proto_values = [
-                python_values_to_proto_values(
-                    transformed_features_df[feature].values, ValueType.UNKNOWN
+            proto_values = []
+            for feature in selected_subset:
+                feature_vector = transformed_features[feature]
+                proto_values.append(
+                    python_values_to_proto_values(feature_vector, ValueType.UNKNOWN)
```

Comment:
**Ross**: How does this work? Don't we still need to extract .values from the returned pd.Series in the Pandas case?